### PR TITLE
Add GitHub Actions runner application

### PR DESCRIPTION
**Pull request by dhirschfeld**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by dhirschfeld** [comment]
```
cd ./src
dotnet msbuild -t:Build -p:PackageRuntime="linux-x64" -p:BUILDCONFIG="Release" -p:RunnerVersion="2.294.0" ./dir.proj
cp -R ./Misc/layoutroot/* ../_layout/
rm -f ../_layout/*.cmd
cp -R ./Misc/layoutbin/* ../_layout/bin/
```
***xref:*** https://github.com/actions/runner/blob/72e2107b5e50a1c1e2b7555215e5829f615f5db5/src/dir.proj#L55-L69

**Comment by dhirschfeld** [comment]
Looks like we need to use the source package as all the others are pre-compiled:
* https://github.com/actions/runner/archive/refs/tags/v2.294.0.tar.gz
* https://github.com/actions/runner/releases/tag/v2.294.0

**Comment by dhirschfeld** [comment]
`git update-index --assume-unchanged` is failing :/
```
  Building : 5ae3e36d71d77189d1a92962facbafa2a62907d4 --- linux-x64
  fatal: Unable to mark file build_artifacts/actions-runner_1658927901239/work/src/Runner.Sdk/BuildConstants.cs
$SRC_DIR/src/dir.proj(30,9): error MSB3073: The command "git update-index --assume-unchanged $SRC_DIR/src/Runner.Sdk/BuildConstants.cs" exited with code 128.
```

**Comment by dhirschfeld** [comment]
```
+ actions-runner config --version
Libicu's dependencies is missing for Dotnet Core 6.0
Execute sudo ./bin/installdependencies.sh to install any missing Dotnet Core 6.0 dependencies.
Tests failed for actions-runner-2.294.0-haf78b9f_0.tar.bz2 -
```
* https://github.com/actions/runner/blob/ca2b1bc6d5eb87339966571a8ad453adcdca4cf2/docs/start/envlinux.md#full-dependencies-list

**Comment by dhirschfeld** [comment]
I think the failing ICU dependencies was a red-herring and the check in the script didn't account for a conda-forge installed ICU package

**Comment by dhirschfeld** [comment]
Yup, works fine without the incorrect check
```
+ actions-runner config --version
2.294.0
+ exit 0
```

**Comment by dhirschfeld** [comment]
This PR builds the GitHub Actions Runner from source. The test passes so I guess this is good for review!

**c.c.** @conda-forge/staged-recipes

**Comment by dhirschfeld** [comment]
Hmm, seems like unvendoring broke things:
```
An error occurred trying to start process '/opt/micromamba/envs/bin/lib/actions-runner/externals/node12/bin/node'. No such file or directory
```
...furthermore, they depend on unsupported node versions with critical CVEs! 🤮 
https://github.com/actions/runner/issues/1886#issuecomment-1136679556

At least there's a workaround:
https://github.com/actions/runner/issues/1911

Will have to try and patch it to use `conda-forge` node...

**Comment by dhirschfeld** [comment]
Have confirmed that the runner works with the unbundled node. I think this is good to merge!

**Review by dhirschfeld — COMMENTED** [review]
This is ready for review/merge!

**Comment by dhirschfeld** [comment]
bump!

**Comment by dhirschfeld** [comment]
Maybe no one is feeling confident reviewing a dotnet PR? 🤔 

The test passes:
```
+ actions-runner config --version
2.294.0
+ exit 0
```
...and I'll be dogfooding it so will be happy to fix any bugs that crop up. I already found and fixed the issue with un-vendoring node.

In the compiled project [released by Microsoft](https://github.com/actions/runner/releases) the only 3rd party licenses are for the vendored node so I think we're good in that regard since we rely on the conda-forge `nodejs` now.

**Comment by dhirschfeld** [comment]
Ping @conda-forge/staged-recipes

**Comment by dhirschfeld** [comment]
Maybe someone from @conda-forge/core could help get this over the line?

**Review by beckermr — CHANGES_REQUESTED** [review]
It looks like the activation scripts will leave empty env vars around in some cases. We should clean up the logic there.

File: recipes/actions-runner/deactivate.d/actions-runner-force-node16.sh [hunk 1]
```diff
@@ -0,0 +1,3 @@
+unset ACTIONS_RUNNER_FORCE_ACTIONS_NODE_VERSION
+
+export ACTIONS_RUNNER_FORCE_ACTIONS_NODE_VERSION=$_ACTIONS_RUNNER_FORCE_ACTIONS_NODE_VERSION
```
Comment (dhirschfeld) [inline recipes/actions-runner/deactivate.d/actions-runner-force-node16.sh:3]: I thought this should be sufficient to cleanup the env vars? Will have to do some testing 🤔

**Comment by dhirschfeld** [comment]
Setting seems to work but can't `deactivate` to test unsetting:
```
+ actions-runner config --version
2.294.0
+ test -n set
+ conda deactivate

CommandNotFoundError: Your shell has not been properly configured to use 'conda deactivate'.
To initialize your shell, run

    $ conda init <SHELL_NAME>

Currently supported shells are:
  - bash
  - fish
  - tcsh
  - xonsh
  - zsh
  - powershell

See 'conda init --help' for more information and options.

IMPORTANT: You may need to close and restart your shell after running 'conda init'.
```

File: recipes/actions-runner/deactivate.d/actions-runner-force-node16.sh (hunk 1 above)
Comment (beckermr) [reply recipes/actions-runner/deactivate.d/actions-runner-force-node16.sh]: What if I don't have this env set initially? Then the env ends up with an empty variable it didn't have iiuic.

**Comment by beckermr** [comment]
Check the compiler packages for how to handle env vars correctly.

**Comment by dhirschfeld** [comment]
Thanks for the review @beckermr!

With this latest commit I think it's checking properly that the variable exists in the activated environment and no longer exists after sourcing the deactivate script:
```
+ actions-runner config --version
2.294.0
+ test -n set
+ source /home/conda/staged-recipes/build_artifacts/actions-runner_1659661275344/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/etc/conda/deactivate.d/actions-runner-force-node16.sh
++ unset ACTIONS_RUNNER_FORCE_ACTIONS_NODE_VERSION
++ '[' -n '' ']'
+ test -z ''
+ exit 0
``` 

Let me know if you're happy with that?

**Comment by dhirschfeld** [comment]
Thanks Matt! 🚀🎉

**Pull request merged**